src/lib.rs

```diff
@@ -285,179 +302,31 @@ fn run_commands<C: Command>(runner: &Runner<C>) -> CommandHandle {
-            message: OutputMessagePayload::Start,
-        });
-
-        let mut cmd_handle = command_process
-            .spawn()
-            .unwrap_or_else(|_| panic!("Unable to spawn process: {}", command.command.clone()));
-        let std_out = cmd_handle.stdout.take();
-        let std_err = cmd_handle.stderr.take();
-        let mut std_out_handle = None;
-        let mut std_err_handle = None;
-
-        let shared_handle = Arc::new(Mutex::new(cmd_handle));
-
-        let child_clone = shared_handle.clone();
-        let kill_trigger_clone = kill_trigger.clone();
-        thread::spawn(move || kill_thread(&kill_trigger_clone, child_clone));
-
-        if let Some(output) = std_out {
-            let mut buffered_stdout = BufReader::new(output);
-            let new_name = command_name.clone();
-            let new_chan = send_chan.clone();
-            std_out_handle = Some(thread::spawn(move || {
-                read_stream(&new_name, new_chan, &mut buffered_stdout, true);
-            }));
-        }
-
-        if let Some(output) = std_err {
-            let mut buffered_stdout = BufReader::new(output);
-            let new_name = command_name.clone();
-            let new_chan = send_chan.clone();
-            std_err_handle = Some(thread::spawn(move || {
-                read_stream(&new_name, new_chan, &mut buffered_stdout, false);
-            }));
-        }
-
-        if let Some(handle) = std_out_handle {
-            let _ = handle.join();
-        }
-
-        if let Some(handle) = std_err_handle {
-            let _ = handle.join();
-        }
-
-        let exit_status = shared_handle.lock().unwrap().wait();
-        match exit_status {
-            Ok(status) => {
-                let _ = send_chan.send(OutputMessage {
-                    name: command_name.clone(),
-                    message: OutputMessagePayload::Done(status.code()),
-                });
-
-                match options.restart {
-                    RestartOptions::Continue => {
-                        return (command_name, Some(status));
-                    }
-                    RestartOptions::Restart => {
-                        if status.success() {
-                            return (command_name, Some(status));
-                        }
-                    }
-                    RestartOptions::Kill => {
-                        if !status.success() {
-                            let _ = kill_trigger.initiate_kill();
-                        }
-                        return (command_name, Some(status));
-                    }
-                };
-            }
-            Err(e) => {
-                let _ = send_chan.send(OutputMessage {
-                    name: command_name.clone(),
-                    message: OutputMessagePayload::Error(e),
-                });
-                return (command_name, None);
-            }
-        }
-    })
+fn check_command(exec_name: &str) -> Result<(), CommandError> {
```

Comment:
Have you tried using the result of Command#spawn? It contains detailed information on why the call failed:

```rs
let mut cmd = std::process::Command::new("zonk");
let res = cmd.spawn();
if let Err(res) = res {
    match res.kind() {
        std::io::ErrorKind::NotFound => todo!(),
        std::io::ErrorKind::PermissionDenied => todo!(),
        _ => todo!(),
    }
}

Reply:
This would save you the extra filesystem operations of the `which` crate.

Reply:
And also give other nice error messages like "permission denied".

Reply:
remember that command construction happens before thread::spawn is called. 
If we want CommandError to return command not found, we can't call thread.spawn. 

CommandError is returned in the Runner/builder. 
However, thread::spawn is only called when that builder actually runs which is a bit later.

We have two choices. Check on build time if the exe exists (current implementation)
Wait for thread::spawn to fail and then return the Error via channel/stdout that command wasn't found if we want to leverage thread::spawn error kinds.

Reply:
Good point. Maybe indicate this error in the result from `execute` or `join` (`ExitResult`)?

What is the correct behavior if the caller starts two processes and process A cannot start (because it doesn't exist or the user doesn't have permissions)? 

It doesn't make sense to try to restart process A. It also doesn't make sense to keep process B running in this case. I think it should kill all other subprocesses and return an error message describing the issue. This could happen before or during `join`.

Currant already has machinery to detect and signal if a command exits. Maybe this machinery can be used to signal if a command couldn't even be started?